HIP/hip-859.md

```diff
@@ -0,0 +1,158 @@
+
+If we sum up all of the `gasUsed` values of each ContractActions sidecar for a given transaction on top of the initial intrinsic gas, which is the initial gas, charged prior to the transaction execution, then we would get accurate value for the `gasConsumed` field.
+
+## User stories
+
+1. As a user, I want to see how much gas was consumed by the EVM during the execution of my transaction, when I retrieve a transaction info from the mirror-node REST APIs, so that I know if I can optimise my transaction cost
+2. As a user, I want to see and compare how much gas I was charged for and how much gas was consumed by the EVM, when I inspect my transaction in hashscan
+
+## Specification
+
+In order to calculate the `gasUsed` field, the mirror node should have `CONTRACT_ACTION` and `CONTRACT_BYTECODE` sidecars enabled. Then, we would use an enhancement logic in the `mirror-node-importer`,
+which will iterate over all `ContractAction` sidecar records for the transaction we want to import and calculate the `gasConsumed` field for.
+
+The new logic would have a mechanism to first calculate the intrinsic gas cost for the transaction being imported. We can inspect if the transaction has a `ContractByteCode` sidecar
+related to it. If yes, we have a contract creation transaction and we would calculate the intrinsic gas cost based on the length of the init bytecode. If we don't have such a sidecar,
+then the transaction represents a contract call and we would use the default intrinsic gas cost for contract calls, which is 21 000.
+
+On top of the intrinsic gas cost, we would iterate over all `ContractAction`
+sidecars and sum up the `gasUsed` field for each of them. The result would be the `gasConsumed` field for the transaction.
+
+```
+Example:
+
+1. The transaction is a ContractCall
+
+gas_consumed = intrinsic_gas + sum(contract_action_gas_used)
+
+2. The transaction is a ContractCreate
+
+gas_consumed = intrinsic_gas + cost(contract_init_bytecode) + sum(contract_action_gas_used)
+```
+
+The `gas_used` field in `contract_result` table would still have the same semantics and represent the gas that was charged to the user.
+
+The following REST APIs will be extended with the new field:
+
+```
+1. GET /api/v1/contracts/results
+2. GET /api/v1/contracts/results/{transactionIdOrHash}
+3. GET /api/v1/contracts/{contractIdOrAddress}/results 
+4. GET /api/v1/contracts/{contractIdOrAddress}/results/{timestamp}
+```
+
+The current response from these APIs looks like this:
+
+```json
+{
+  ...
+  "error_message": "Out of gas",
+  "failed_initcode": "0x856739",
+  "from": "0000000000000000000000000000000000001f41",
+  "function_parameters": "0xbb9f02dc6f0e3289f57a1f33b71c73aa8548ab8b",
+  "gas_limit": 100000,
+  "gas_price": "0x4a817c800",
+  "gas_used": 320000,
+  "hash": "0x3531396130303866616264653464",
+  ...
+}
+```
+
+After the change, it will look like this:
```

Comment:
Don't think you need to show the before and after. We don't normally do that for protos or APIs in HIPs. Can just say explicitly that a `gas_consumed` will be added to these APIs.